### ISSUE TITLE
Multi-GPU data shuffling

### ISSUE THREAD

**Issue by cxmzdd** [OpenHLM-project/OpenHLM#3]
Hi, thanks for open-sourcing OpenHLM! I have a question about  Multi-GPU data shuffling in `scripts/train_pytorch.py`.
In the PyTorch training loop, there is code intended to update the sampler epoch for distributed training:
https://github.com/OpenHLM-project/OpenHLM/blob/4a8d4cd0ecd8313bb8dbb10d2afac2c2ec9ed1b4/src/openpi4OpenHLM/scripts/train_pytorch.py#L1135-L1136
However, loader is created via create_data_loader() and returned as DataLoaderImpl. Neither DataLoaderImpl nor TorchDataLoader defines a set_epoch() method. For PyTorch DistributedSampler, set_epoch(epoch) is required to get a different shuffle order on each pass over the dataset. Without it, each epoch may use the same index ordering. 
Is this a bug?

**Comment by Tendourisu** [comment]
Thank you for identifying and reporting this bug!

We checked the relevant code path and confirmed that `DataLoaderImpl` and `TorchDataLoader` do not expose `set_epoch()`, so the following condition never succeeds:

```python
if use_ddp and hasattr(loader, "set_epoch"):
    loader.set_epoch(global_step // len(loader))
```

The same issue is also present in the upstream Physical Intelligence `openpi` repository:

- The training loop contains the same guarded `set_epoch()` call:  
  https://github.com/Physical-Intelligence/openpi/blob/15a9616a00943ada6c20a0f158e3adb39df2ccac/scripts/train_pytorch.py#L510-L512
- A `DistributedSampler` is created and passed to `TorchDataLoader` here:  
  https://github.com/Physical-Intelligence/openpi/blob/15a9616a00943ada6c20a0f158e3adb39df2ccac/src/openpi/training/data_loader.py#L305-L337
- `TorchDataLoader` does not implement `set_epoch()` or `__len__()`:  
  https://github.com/Physical-Intelligence/openpi/blob/15a9616a00943ada6c20a0f158e3adb39df2ccac/src/openpi/training/data_loader.py#L381-L468
- The outer `DataLoaderImpl` wrapper also does not implement either method:  
  https://github.com/Physical-Intelligence/openpi/blob/15a9616a00943ada6c20a0f158e3adb39df2ccac/src/openpi/training/data_loader.py#L530-L540

Therefore, this appears to be an issue inherited from the upstream implementation rather than something introduced specifically by OpenHLM.

For our released setting, however, we expect the practical impact to be small. The `g1_HLM-12_full_teleop` dataset contains 522,181 frames, and our 10k-step setting with a global batch size of 128 corresponds to approximately 2.45 passes over the dataset. This means that only one complete sample ordering is repeated, followed by roughly 45% of a third pass.

The initial ordering is still randomly shuffled, and the model independently resamples image augmentations, flow-matching noise, and timesteps during training. Consequently, repeated sample ordering does not produce identical training inputs or gradients.

Given these factors, we do not expect this bug to have a material effect on the released checkpoint, and we consider it safe to ignore for this particular training setting.

**Comment by cxmzdd** [comment]
Thank you for the reply.
